lib/nameMatcher.js

```diff
@@ -0,0 +1,307 @@
+    var lastNamePromise = promisify(
+            redisClient.hset,
+            [ ((environment === "prod") ? 'lastNames' : 'lastNamesDev'), last, lastMPString ],
+            redisClient)
+            .catch(function catchAddLastNameError(err) {
+                console.log('Error adding last name. Reverting value');
+                // Restore localCollection to original value
+                lastNames[last] = backupLastName;
+                throw (err);
+            });
+    return Promise.all([lastNamePromise, firstNamePromise])
+        .then(function logAddNameSuccess(nameResults) {
+            console.log('nameMatcher successfully added name');
+            return true;
+        })
+        .catch(function catchAddNameError(err) {
+            console.error('Encountered error refreshing local names');
+            console.error(err);
+            throw err;
+        });
+};
+
+
+/**
+ * match the input name against names in firstNames and lastNames
+ * @param name {object} an object with a 'first' and 'last' property
+ * @return {object} an object with the following properties:
+ * original: the original name object provided as input
+ * first: an array of first names that match original.first
+ * last: an array of last names that match original.last
+ */
+nameMatcher.matchName = function (name, environment) {
+    var matches = { original: name, first: [], last: [] };
+    if (environment === "prod") {
+        matches.first = _.keys(firstNames)
+            .filter(function compareFirstNames(localName) {
+                return compareNames(name.first, localName, firstNames);
+        });
+        matches.last = _.keys(lastNames)
+            .filter(function compareLastNames(localName) {
+                return compareNames(name.last, localName, lastNames);
+        });
+    } else if (environment === "dev") {
+        matches.first = _.keys(firstNamesDev)
+            .filter(function compareFirstNames(localName) {
+                return compareNames(name.first, localName, firstNamesDev);
+        });
+        matches.last = _.keys(lastNamesDev)
+            .filter(function compareLastNames(localName) {
+                return compareNames(name.last, localName, lastNamesDev);
+        });
+    }
+    return matches;
+};
+
+
+/**
+ * calls matchName for each name in nameList
+ * @param nameList {array} an array of objects with a 'first' and 'last' property
+ * @return {array} an array of objects with the following properties:
+ * original: the original name object provided as input
+ * first: an array of first names that match original.first
+ * last: an array of last names that match original.last
+ */
+nameMatcher.matchNames = function (nameList, environment) {
+    var tempMatcher = this;
+    return nameList.map(function(name) { 
+        return tempMatcher.matchName(name, environment); 
+    });
+};
+
+
+/**
+ * close the redis client connection
+ */
+nameMatcher.close = function () {
+    redisClient.quit();
```

Comment:
It does, it's pretty cool. Redis keeps everything in memory and when it's not busy a background process writes it to disk